### PR TITLE
Implement event engine cucumber tests

### PR DESCRIPTION
**Pull request by kleewho**
Subscribe remoteAction can't be reusable
Move event engine specific contract test stuff to event engine package 
MappingRemoteAction doesn't require result in constructor

### CODE REVIEW

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/state/SubscribeState.kt [hunk 1]
```diff
@@ -101,8 +101,20 @@ sealed class SubscribeState : State<SubscribeEffectInvocation, Event, SubscribeS
...
                     transitionTo(HandshakeStopped(channels, channelGroups, reason))
                 }
-                is Event.HandshakeReconnectGiveUp -> {
-                    transitionTo(HandshakeFailed(this.channels, this.channelGroups, event.reason))
+                is Event.HandshakeReconnectGiveup -> {
+                    transitionTo(
+                        HandshakeFailed(this.channels, this.channelGroups, event.reason),
+                        SubscribeEffectInvocation.EmitStatus(
+                            PNStatus(
+                                category = PNStatusCategory.PNReconnectionAttemptsExhausted,
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/subscribe/eventengine/state/SubscribeState.kt:109]: This depends on what we decide to return

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt [hunk 2]
```diff
@@ -24,13 +24,13 @@ sealed class Event : CoreEvent {
     data class HandshakeReconnectSuccess(val subscriptionCursor: SubscriptionCursor) : Event()
     data class HandshakeReconnectFailure(val reason: PubNubException) : Event()
     object HandshakeReconnectRetry : Event()
-    data class HandshakeReconnectGiveUp(val reason: PubNubException) : Event()
+    data class HandshakeReconnectGiveup(val reason: PubNubException) : Event()
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt:27]: Why this change?

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt [hunk 3]
```diff
@@ -24,13 +24,13 @@ sealed class Event : CoreEvent {
...
 
     data class ReceiveSuccess(val messages: List<PNEvent>, val subscriptionCursor: SubscriptionCursor) : Event()
     data class ReceiveFailure(val reason: PubNubException) : Event()
     data class ReceiveReconnectSuccess(val messages: List<PNEvent>, val subscriptionCursor: SubscriptionCursor) : Event()
     data class ReceiveReconnectFailure(val reason: PubNubException) : Event()
     object ReceiveReconnectRetry : Event()
-    data class ReceiveReconnectGiveUp(val reason: PubNubException) : Event()
+    data class ReceiveReconnectGiveup(val reason: PubNubException) : Event()
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt:34]: Why this change?

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt (hunk 2 above)
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt:27]: This is how we named it in the spec https://github.com/pubnub/ce-decisions/blob/master/documents/Subscribe%20Event%20Engine%20Description.md#handshake_reconnect_giveup and it's how it's used in the contract tests https://github.com/pubnub/sdk-specifications/blob/001a288adefa3ac2ba1f02d83e5e8c955b0f4349/features/subscribe/event-engine/happy-path.feature#L47

File: src/main/kotlin/com/pubnub/api/PubNub.kt [hunk 4]
```diff
@@ -86,15 +87,21 @@ import com.pubnub.api.models.consumer.objects.membership.PNChannelDetailsLevel
...
+import com.pubnub.api.workers.SubscribeMessageProcessor
 import java.io.InputStream
 import java.util.Date
 import java.util.UUID
 
-class PubNub internal constructor(val configuration: PNConfiguration, eventEngineConf: EventEngineConf) {
+class PubNub internal constructor(
+    val configuration: PNConfiguration,
+    eventEngineConf: EventEngineConf,
+    retryPolicy: RetryPolicy = configuration.retryPolicy()
```
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:103]: To run contract tests in a timely manner I need to be able to configure the delay between retries. We do not create Subscribe in the tests and due to circular dependency it's not possible to pass it to PubNub instance so it seemed like this is the only option available.

File: src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt [hunk 5]
```diff
@@ -88,6 +86,7 @@ class Subscribe(
                     maxRetries = pubNub.configuration.maximumReconnectionRetries,
                     fixedDelay = Duration.ofSeconds(3)
                 )
+
                 PNReconnectionPolicy.EXPONENTIAL -> ExponentialPolicy(maxRetries = pubNub.configuration.maximumReconnectionRetries)
```
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt:90]: No. Not used, not needed

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt [hunk 6]
```diff
@@ -1,19 +1,22 @@
...
         channels: List<String>,
         channelGroups: List<String>
     ): RemoteAction<SubscriptionCursor> {
+        val subscribe = Subscribe(pubNub)
         subscribe.channels = channels
         subscribe.channelGroups = channelGroups
         subscribe.timetoken = 0
         subscribe.region = null
-        return RemoteActionForHandshake(subscribe)
+        return subscribe.map { SubscriptionCursor(it.metadata.timetoken, it.metadata.region) }
```
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt:20]: Is there anything that could make it more readable?

File: src/main/kotlin/com/pubnub/api/PubNub.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:103]: My understanding is that in test you can set retryPolicy in configuration and then create PubNub instance containing desired retryPolicy. Is this not enough for test to work properly?

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt (hunk 2 above)
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt:27]: Ok, thank's for clarification.

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt (hunk 3 above)
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/event/Event.kt:34]: I see.

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt:20]: For me those separate classes like RemoteActionForReceiveMessagesResult.kt clearly stated what code is doing but that maybe is related to my lack of experience in such abstractions.

What might improve code in the current shape is renam of `map` class to `mapTo..... `This is because map might be confuse with standard kotlin `map` function

File: src/main/kotlin/com/pubnub/api/PubNub.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:103]: It's not enough. We can't change the delay in configuration and without changing the delay these tests would take too much time

[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:103]: > My understanding is that in test you can set retryPolicy in configuration 

We don't have such field in configuration

[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:103]: So maybe adding `delay` as internal parameter in config would solve this inconvenience.  My point is that adding retryPolicy as parameter to PubNub constructor is something that I would try to avoid.

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt:20]: e.g. mapToExtendedRemoteAction.

[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt:20]: Will `mapResult` be fine with you?

[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt:20]: Choosing between `map` and `mapResult` I would go with map.

[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/effectprovider/HandshakeProviderImpl.kt:20]: `map` that is

**Review by marcin-cebo — APPROVED** [review]
I appreciate that PubNub constructor didn't get bigger.
